### PR TITLE
KYLIN-4857 Refactor system cube for kylin4

### PR DESCRIPTION
**Pull request by zhangayqian**
## Proposed changes

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce to Kylin?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have create an issue on [Kylin's jira](https://issues.apache.org/jira/browse/KYLIN), and have described the bug/feature there in detail
- [x] Commit messages in my PR start with the related jira ID, like "KYLIN-0000 Make Kylin project open-source"
- [ ] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If this change need a document change, I will prepare another pr against the `document` branch
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at user@kylin or dev@kylin by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -1465,6 +1465,7 @@
         <artifactId>scalastyle-maven-plugin</artifactId>
         <version>1.0.0</version>
         <configuration>
+          <skip>true</skip>
```
Comment (hit-lacus) [inline pom.xml:1468]: Remove this line.

File: core-metrics/src/main/java/org/apache/kylin/metrics/QuerySparkMetrics.java [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentMap;
+import java.util.stream.Collectors;
+
+public class QuerySparkMetrics {
+    private static Map<String, List<SparkJobMetrics>> sparkJobMetricsMap = Maps.newConcurrentMap();
+    private static List<SparkStageMetrics> sparkStageMetricsList = new ArrayList<>();
```
Comment (hit-lacus) [inline core-metrics/src/main/java/org/apache/kylin/metrics/QuerySparkMetrics.java:14]: Use a static collection is not a good practice . Maybe using a `BlockingQueue` is better, like [Producer-Consumer pattern](https://blog.csdn.net/qq_35580883/article/details/78650677) , or send directly to `AbstractActiveReservoir` is also good to me.

**Comment by hit-lacus** [comment]
Following is my suggestion:

You are using event flow : SparderMetricsListener  -> QueryService -> MetricsSystem, I think it is OK if you send MetricsEvent direct to MetricsSystem : SparderMetricsListener  -> MetricsSystem.

In the previous implementation, when a query is finished, all metrics data are collected because query and metrics collection are in the same thread. But in the new implementation, SparkListener and QueryService are not in the same thread, so the order is not guaranteed .

File: core-metrics/src/main/java/org/apache/kylin/metrics/QuerySparkMetrics.java [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+    private static Map<Long, List<SparkStageMetrics>> sparkStageMetricsMap = Maps.newConcurrentMap();
+    private static List<SparkJobMetrics> sparkJobMetricsList = new ArrayList<>();
+    private static Map<String, QueryExecutionMetrics> queryExecutionMetricsMap = Maps.newConcurrentMap();
+
+    public static void addSparkStageMetrics(int jobId, int stageId, String stageType, boolean isSuccess, long resultSize,
```
Comment (hit-lacus) [inline core-metrics/src/main/java/org/apache/kylin/metrics/QuerySparkMetrics.java:19]: It is better to use a **Builder** Pattern, like a `SparkStageMetricsBuilder` .

File: core-metrics/src/main/java/org/apache/kylin/metrics/property/QuerySparkStageEnum.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+    private final String propertyName;
+
+    QuerySparkStageEnum(String name) {
+        this.propertyName = name;
+    }
+
+    public static QuerySparkStageEnum getByName(String name) {
+        if (Strings.isNullOrEmpty(name)) {
+            return null;
```
Comment (hit-lacus) [inline core-metrics/src/main/java/org/apache/kylin/metrics/property/QuerySparkStageEnum.java:36]: I guess it is better to throw a IllegalArgumentException.

File: core-metrics/src/main/java/org/apache/kylin/metrics/QuerySparkMetrics.java [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+import java.util.concurrent.TimeUnit;
+
+public class QuerySparkMetrics {
+    private static final Logger logger = LoggerFactory.getLogger(QuerySparkMetrics.class);
+    private static final QuerySparkMetrics instance = new QuerySparkMetrics();
+    private org.apache.kylin.shaded.com.google.common.cache.Cache<String, QueryExecutionMetrics> queryExecutionMetricsMap;
+
+    private QuerySparkMetrics() {
+        queryExecutionMetricsMap = CacheBuilder.newBuilder()
+                .maximumSize(1000)
```
Comment (hit-lacus) [inline core-metrics/src/main/java/org/apache/kylin/metrics/QuerySparkMetrics.java:22]: Make 1000 configurable .

File: core-metrics/src/main/java/org/apache/kylin/metrics/QuerySparkMetrics.java [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+                .expireAfterWrite(600, TimeUnit.SECONDS)
```
Comment (hit-lacus) [inline core-metrics/src/main/java/org/apache/kylin/metrics/QuerySparkMetrics.java:23]: Make 600 configurable .

File: kylin-spark-project/kylin-spark-query/src/main/scala/org/apache/spark/sql/metrics/SparderMetricsListener.scala [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+  val queryExecutionMetrics = QuerySparkMetrics.getInstance()
+
+  override def onJobStart(event: SparkListenerJobStart): Unit = {
+    val executionIdString = event.properties.getProperty(SQLExecution.EXECUTION_ID_KEY)
+    val sparderName = event.properties.getProperty("spark.app.name")
+    val kylinQueryId = event.properties.getProperty("kylin.query.id")
+
+    if (executionIdString == null || kylinQueryId == null) {
+      // This is not a job created by SQL
+      logInfo("Cannot happened.")
```
Comment (hit-lacus) [inline kylin-spark-project/kylin-spark-query/src/main/scala/org/apache/spark/sql/metrics/SparderMetricsListener.scala:41]: Remove this line.

File: core-metrics/src/main/java/org/apache/kylin/metrics/property/QuerySparkExecutionEnum.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    PROJECT("PROJECT"),
+    TYPE("QUERY_TYPE"),
+    REALIZATION("REALIZATION"),
+    REALIZATION_TYPE("REALIZATION_TYPE"),
+    QUERY_ID("QUERY_ID"),
+    EXECUTION_ID("EXECUTION_ID"),
+    USER("KUSER"),
+    SPARDER_NAME("SPARDER_NAME"),
+    EXCEPTION("EXCEPTION"),
+    STRAT_TIME("STRAT_TIME"),
```
Comment (hit-lacus) [inline core-metrics/src/main/java/org/apache/kylin/metrics/property/QuerySparkExecutionEnum.java:17]: Typo!

File: core-metrics/src/main/java/org/apache/kylin/metrics/property/QuerySparkExecutionEnum.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
+package org.apache.kylin.metrics.property;
```
Comment (zzcclp) [inline core-metrics/src/main/java/org/apache/kylin/metrics/property/QuerySparkExecutionEnum.java:1]: Add license header

File: core-metrics/src/main/java/org/apache/kylin/metrics/property/QuerySparkStageEnum.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
+package org.apache.kylin.metrics.property;
```
Comment (zzcclp) [inline core-metrics/src/main/java/org/apache/kylin/metrics/property/QuerySparkStageEnum.java:1]: Add license header

File: core-metrics/src/main/java/org/apache/kylin/metrics/property/QuerySparkJobEnum.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
+package org.apache.kylin.metrics.property;
```
Comment (zzcclp) [inline core-metrics/src/main/java/org/apache/kylin/metrics/property/QuerySparkJobEnum.java:1]: Add license header

File: core-metrics/src/main/java/org/apache/kylin/metrics/lib/impl/MetricsSystem.java [hunk 12]
```diff
@@ -26,6 +26,7 @@
 import java.util.concurrent.ConcurrentHashMap;
 
 import org.apache.kylin.common.KylinConfig;
+import org.apache.kylin.metrics.QuerySparkMetrics;
```
Comment (zzcclp) [inline core-metrics/src/main/java/org/apache/kylin/metrics/lib/impl/MetricsSystem.java:29]: remove this unused import

File: core-metrics/src/main/java/org/apache/kylin/metrics/QuerySparkMetrics.java [hunk 13]
```diff
@@ -0,0 +1,398 @@
+package org.apache.kylin.metrics;
```
Comment (zzcclp) [inline core-metrics/src/main/java/org/apache/kylin/metrics/QuerySparkMetrics.java:1]: Add license header

File: core-metrics/src/main/java/org/apache/kylin/metrics/QuerySparkMetrics.java [hunk 14]
```diff
@@ -0,0 +1,398 @@
...
+        sparkStageMetrics.setStageType(stageType);
+        sparkStageMetrics.setStartTime(startTime);
+        queryExecutionMetricsMap.asMap().get(queryId).getSparkJobMetricsMap().get(jobId).getSparkStageMetricsMap().put(stageId, sparkStageMetrics);
+
+    }
+
+    public void updateSparkStageMetrics(String queryId, int jobId, int stageId, boolean isSuccess, long resultSize,
+                                     long executorDeserializeTime, long executorDeserializeCpuTime, long executorRunTime,
+                                     long executorCpuTime, long jvmGCTime, long resultSerializationTime, long memoryBytesSpilled,
+                                     long diskBytesSpilled, long peakExecutionMemory) {
```
Comment (zzcclp) [inline core-metrics/src/main/java/org/apache/kylin/metrics/QuerySparkMetrics.java:75]: There are too many parameters, it's more than 12, consider to use a class to indicate.

**Comment by hit-lacus** [comment]
Thanks @zhangayqian for your great works, and thanks @zzcclp for review and verification, both you are excellect !
This patch looks good to me, I will merge this.

**Review by hit-lacus — APPROVED** [review]
Approved

**Pull request merged**